File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-16.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        def isSameTree(root, sub):
            if not root and not sub:
                return True

            if root and sub and root.val == sub.val:
                return isSameTree(root.left, sub.left) and isSameTree(root.right, sub.right)

            else:
                return False

        if not root:
            return False

        if not subRoot:
            return True

        if isSameTree(root, subRoot):
            return True

        return self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-16.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int goodNodes(TreeNode* root) {
        if (!root) {
            return 0;
        }

        std::queue<pair<TreeNode*, int>> q;
        q.push({root, INT_MIN});
        int res = 0;

        while (!q.empty()) {
            TreeNode* node = q.front().first;
            int maxVal = q.front().second;
            q.pop();

            if (node->val >= maxVal) {
                res++;
            }

            maxVal = max(node->val, maxVal);
            if (node->left != nullptr) {
                q.push({node->left, maxVal});
            }

            if (node->right != nullptr) {
                q.push({node->right, maxVal});
            }
        }

        return res;
    }
};
